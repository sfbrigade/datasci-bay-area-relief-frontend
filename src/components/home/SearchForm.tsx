import styled from "styled-components";
import Select from "@mui/material/Select";
import {County, OrgType} from "../../types";
import FormControl from "@mui/material/FormControl";
import InputLabel from "@mui/material/InputLabel";
import Button from "@mui/material/Button";
import React, {useState} from "react";
import {getFilterNameFromGroupAndTargetName} from "../results/filterHelpers";
import {useNavigate} from "react-router-dom";
import {Typography} from "@mui/material";
import {colors} from "../../theme";

const SearchFormContainer = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 14px;
  border: 0;
`;

const SearchDescription = styled.div`
  max-width: 65%;
  z-index: 10;
  margin-bottom: 27px;
`;

const SearchFormFields = styled.div`
  display: flex;
  flex-direction: column;
  width: 328px;
`;

const StyledSelect = styled(Select)`
  margin-bottom: 1em;
`;

const SearchButton = styled(Button)`
  && {
    width: 97px;
    height: 36px;
    border-radius: 200px;

    :disabled {
      background-color: ${colors.primaryRed};
      color: black;
      opacity: 0.5;
    }
  }
`;
const DonateButton = styled(Button)`
  && {
    margin-top: 13px;
    padding: 13px;
    border-radius: 4px;
  }
`;

export const SearchForm = () => {
  const [orgType, setOrgType] = useState<OrgType>(OrgType.Any);
  const [county, setCounty] = useState<County>(County.Any);

  const navigate = useNavigate();

  const goToResults = () => {
    let path = "/results";
    if (orgType !== "any") {
      const orgTypeParam = getFilterNameFromGroupAndTargetName(
        "orgType",
        orgType
      );
      //append orgType=blah
      path = `${path}?orgType=${orgTypeParam}&`;
    }
    //check if county is NOT any, append county
    const countyParam = getFilterNameFromGroupAndTargetName("county", county);

    if (countyParam && countyParam !== "any") {
      if (orgType === "any") {
        path = `${path}?`;
      }
      path = `${path}county=${countyParam}`;
    }
    //if (path[path.length -1] === ? or === &, slice it out
    if (path[path.length - 1] === "?" || path[path.length - 1] === "&") {
      path = path.slice(0, path.length - 1);
    }
    navigate(path);
  };

  const goToDonate = () => {
    navigate("/donate");
  };

  return (
    <SearchFormContainer>
      <SearchDescription>
        <Typography variant="body1">
          Search our database for Bay Area loans for your non-profit or small
          business.
        </Typography>
      </SearchDescription>

      <SearchFormFields>
        <FormControl>
          <InputLabel htmlFor="org-type-select">I am a...</InputLabel>
          <StyledSelect
            native
            value={orgType}
            onChange={(event) => setOrgType(event.target.value as OrgType)}
            label="I am a..."
            inputProps={{
              name: "org-type",
              id: "org-type-select",
              "data-testid": "org-type-select",
            }}
          >
            <option value={OrgType.Any}>Any</option>
            <option value={OrgType.SmallBusiness}>Small business</option>
            <option value={OrgType.NonProfit}>Non-profit</option>
          </StyledSelect>
        </FormControl>
        <FormControl>
          <InputLabel htmlFor="county-select">County</InputLabel>
          <StyledSelect
            native
            value={county}
            onChange={(event) => setCounty(event.target.value as County)}
            label="County"
            inputProps={{
              name: "county",
              id: "county-select",
            }}
          >
            <option value={County.Any}>Any</option>
            <option value={County.SanFrancisco}>San Francisco</option>
            <option value={County.Alameda}>Alameda</option>
            <option value={County.SanMateo}>San Mateo</option>
            <option value={County.ContraCosta}>Conta Costa</option>
            <option value={County.SantaClara}>Santa Clara</option>
          </StyledSelect>
        </FormControl>
      </SearchFormFields>
      <SearchButton
        variant="contained"
        color="primary"
        onClick={goToResults}
        disabled={!orgType || !county}
      >
        Search
      </SearchButton>
      <DonateButton color="primary" onClick={goToDonate}>
        I want to donate
      </DonateButton>
    </SearchFormContainer>
  );
};
