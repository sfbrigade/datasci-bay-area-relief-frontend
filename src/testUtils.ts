import {act} from '@testing-library/react';

export const idleForIO = async (): Promise<void> => {
    await act(async () => {
        await new Promise((resolve) => setImmediate(resolve));
    });
};
