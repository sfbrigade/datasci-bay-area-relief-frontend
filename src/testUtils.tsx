import React from "react";
import {act, render, RenderResult} from "@testing-library/react";
import {createMemoryHistory, MemoryHistory} from "history";
import {Router} from "react-router-dom";

export const idleForIO = async (): Promise<void> => {
  await act(async () => {
    await new Promise((resolve) => resolve());
  });
};

export type RenderResultWithHistory = RenderResult & {history: MemoryHistory};
type RenderWithRouterFunction = (
  ui: React.ReactElement,
  routeOptions?: {path?: string; history?: MemoryHistory}
) => RenderResultWithHistory;

export const renderWithRouter: RenderWithRouterFunction = (
  ui,
  {path = "/", history = createMemoryHistory({initialEntries: [path]})} = {}
) => {
  const Wrapper: React.FC = ({children}) => (
    <Router history={history}>{children}</Router>
  );
  return {
    ...render(ui, {wrapper: Wrapper}),
    history,
  };
};
