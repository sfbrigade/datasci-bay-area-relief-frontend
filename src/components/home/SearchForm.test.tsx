import React from "react";
import {fireEvent, render, screen} from "@testing-library/react";
import {SearchForm} from "./SearchForm";
import {idleForIO} from "../../testUtils";
import {createMemoryHistory} from "history";
import {Route, Router} from "react-router-dom";
import renderer from 'react-test-renderer';

const history = createMemoryHistory();
jest.spyOn(history, "push");

describe('SearchForm selections', () => {
  beforeEach(() => {
    render(
      <Router history={history}>
        <Route component={SearchForm} />
      </Router>
    );
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<SearchForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a SearchForm with two default Anys', () => {
    const anyObject = screen.getAllByRole('option', {name: 'Any'});
    expect(anyObject).toHaveLength(2);
  });

  it('can search with no user selections', async () => {
    const searchButton = screen.getByText("Search");
    fireEvent.click(searchButton);

    expect(history.push).toHaveBeenLastCalledWith("/results");

    const anyObject = screen.getAllByRole('option', {name: 'Any'});
    expect((anyObject[0] as HTMLOptionElement).selected).toBeTruthy();
    expect((anyObject[1] as HTMLOptionElement).selected).toBeTruthy();
  });

  it('can search by orgType', async () => {
    fireEvent.change(screen.getByLabelText('I am a...'),
      { target: { value: 'smallBusiness' } });

    const searchButton = screen.getByText("Search");
    fireEvent.click(searchButton);

    await idleForIO();

    expect(history.push).toHaveBeenLastCalledWith("/results?orgType=smallBusiness");

    const smallBusinessOption = screen.getAllByRole('option', {name: 'Small business'});
    expect((smallBusinessOption[0] as HTMLOptionElement).selected).toBeTruthy();
    const anyObject = screen.getAllByRole('option', {name: 'Any'});
    expect((anyObject[0] as HTMLOptionElement).selected).toBeFalsy();
    expect((anyObject[1] as HTMLOptionElement).selected).toBeTruthy();
  });

  it('can search by county only', async () => {
    fireEvent.change(screen.getByLabelText("County"),
      { target: { value: 'alamedaCounty' } });

    const searchButton = screen.getByText("Search");
    fireEvent.click(searchButton);

    await idleForIO();

    expect(history.push).toHaveBeenLastCalledWith("/results?county=alamedaCounty");

    const countyOption = screen.getAllByRole('option', {name: 'Alameda'});
    expect((countyOption[0] as HTMLOptionElement).selected).toBeTruthy();
    const anyObject = screen.getAllByRole('option', {name: 'Any'});
    expect((anyObject[0] as HTMLOptionElement).selected).toBeTruthy();
    expect((anyObject[1] as HTMLOptionElement).selected).toBeFalsy();
  });

  it('can search by county and org type', async () => {
    fireEvent.change(screen.getByLabelText('I am a...'),
      { target: { value: 'smallBusiness' } });
    fireEvent.change(screen.getByLabelText("County"),
      { target: { value: 'alamedaCounty' } });

    const searchButton = screen.getByText("Search");
    fireEvent.click(searchButton);

    await idleForIO();

    expect(history.push).toHaveBeenLastCalledWith("/results?orgType=smallBusiness&county=alamedaCounty");

    const countyOption = screen.getAllByRole('option', {name: 'Alameda'});
    expect((countyOption[0] as HTMLOptionElement).selected).toBeTruthy();
    const smallBusinessOption = screen.getAllByRole('option', {name: 'Small business'});
    expect((smallBusinessOption[0] as HTMLOptionElement).selected).toBeTruthy();
  });
});
