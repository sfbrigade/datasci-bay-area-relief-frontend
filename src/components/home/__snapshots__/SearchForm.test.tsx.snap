// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm selections renders correctly 1`] = `
<form
  className="sc-bdnylx wxXGl"
>
  <div
    className="sc-gtssRu eAezmj"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      Search our database for Bay Area loans for your non-profit or small business.
    </p>
  </div>
  <div
    className="sc-dlnjPT eHtjKL"
  >
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="org-type-select"
      >
        I am a...
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root sc-hKFyIo kKqkRf MuiInputBase-formControl"
        onClick={[Function]}
      >
        <select
          aria-invalid={false}
          className="MuiSelect-root MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
          data-testid="org-type-select"
          disabled={false}
          id="org-type-select"
          multiple={false}
          name="org-type"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          value="any"
        >
          <option
            value="any"
          >
            Any
          </option>
          <option
            value="smallBusiness"
          >
            Small business
          </option>
          <option
            value="nonProfit"
          >
            Non-profit
          </option>
        </select>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
          >
            <span>
              I am a...
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink={true}
        htmlFor="county-select"
      >
        County
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root sc-hKFyIo kKqkRf MuiInputBase-formControl"
        onClick={[Function]}
      >
        <select
          aria-invalid={false}
          className="MuiSelect-root MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="county-select"
          multiple={false}
          name="county"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          value="any"
        >
          <option
            value="any"
          >
            Any
          </option>
          <option
            value="sfCounty"
          >
            San Francisco
          </option>
          <option
            value="alamedaCounty"
          >
            Alameda
          </option>
          <option
            value="sanMateoCounty"
          >
            San Mateo
          </option>
          <option
            value="contraCostaCounty"
          >
            Conta Costa
          </option>
          <option
            value="santaClaraCounty"
          >
            Santa Clara
          </option>
        </select>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
          >
            <span>
              County
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-eCApGN beDnGC MuiButton-containedPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Search
    </span>
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text sc-jSFkmK giMmbY MuiButton-textPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      I want to donate
    </span>
  </button>
</form>
`;
